### ISSUE TITLE
Parameters that cause non-determinism on GPU

### ISSUE THREAD

**Issue by liberbey** [catboost/catboost#1721]
Hi, I'm using CatBoost on GPU and I would like to learn parameters that are cause of non-determinism. I have made some different trials and it seems that CatBoost is sometimes deterministic depending on the parameters, since it gave the same result for 10 trials with these parameters:

```
"modelParam"             : {'bagging_temperature': 3,
                                                                 'iterations': 600,
                                                                 'l2_leaf_reg': 42,
                                                                 'learning_rate': 0.03,
                                                                 'max_bin': 48,
                                                                 'max_depth': 15,
                                                                 'max_leaves': 100,
                                                                 'min_data_in_leaf': 17500,
                                                                 'score_function': 'L2',
                                                                 'grow_policy': 'Lossguide',
                                                                 'loss_function': 'RMSE',
                                                                 'task_type': 'GPU',
                                                                 "has_time": True,
                                                                 "devices": "0"}
```

But when I use these parameters, the results are different for each trial:

```
"modelParam"             : {'bagging_temperature': 27.0,
                                                                 'iterations': 1000.0,
                                                                 'l2_leaf_reg': 60.0,
                                                                 'learning_rate': 0.36,
                                                                 'max_bin': 42.0,
                                                                 'max_ctr_complexity': 11.0,
                                                                 'max_depth': 16.0,
                                                                 'max_leaves': 40.0,
                                                                 'min_data_in_leaf': 12000.0,
                                                                 'model_size_reg': 0.8,
                                                                 'one_hot_max_size': 2,
                                                                 'score_function': 'L2',
                                                                 'grow_policy': 'Lossguide',
                                                                 'loss_function': 'RMSE',
                                                                 'devices': '0:1:2:3',
                                                                 'task_type': 'GPU',
                                                                 'has_time': True,
                                                                 'cat_features': ['country', 'sub_sector', 'sub_industry', 'industry', 'sector']}
```

Is there a list of parameters which cause non-determinism on GPU? Since my dataset is very large I need to use GPU and reproducability is very important for me. Thanks a lot!

catboost version: 0.25.1
Operating System: ubuntu 20.04
CPU:
GPU: NVIDIA Tesla T4

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Currently, CatBoost on GPU builds floating-point histograms using a non-deterministic algorithm -- this cannot be changed by any training parameters.

However, you may disable the random noise added at during training to increase accuracy of the model.
has_time: True (do not shuffle dataset before training)
bootstrap_type: 'No' (disable bootstrap)
random_strength: 0 (do not add random noise when selecting the best split)

**Comment by liberbey** [comment]
Thanks! In my case, disabling bootstrap worked for me. I guess the differences are so small that does not change anything.
